### PR TITLE
VIH-6235 Allow participants to be transferred in or out of hearing rooms

### PR DESCRIPTION
**Pull request by shaed-parkar**
### JIRA link (if applicable) ###

VIH-6235

### Change description ###

* Allow participants to be transferred in or out of hearing rooms

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by valentinabennett — APPROVED** [review]

File: VideoAPI/Video.API/Controllers/ConferenceManagementController.cs [hunk 1]
```diff
@@ -97,5 +98,45 @@ public async Task<IActionResult> EndVideoHearingAsync(Guid conferenceId)
...
+        [ProducesResponseType((int) HttpStatusCode.OK)]
+        public async Task<IActionResult> TransferParticipantAsync(Guid conferenceId, TransferParticipantRequest transferRequest)
+        {
+            _logger.LogDebug("TransferParticipant");
+            try
+            {
+                switch (transferRequest.TransferType)
+                {
+                    case TransferType.Call:
+                        _logger.LogDebug($"Transferring participant {transferRequest.ParticipantId} into hearing room");
```
Comment (levibutler) [inline VideoAPI/Video.API/Controllers/ConferenceManagementController.cs:119]: log messages could have custom dimensions for the conference id / participant id

File: VideoAPI/Video.API/Controllers/ConferenceManagementController.cs [hunk 2]
```diff
@@ -97,5 +98,45 @@ public async Task<IActionResult> EndVideoHearingAsync(Guid conferenceId)
...
+                            transferRequest.ParticipantId,
+                            RoomType.HearingRoom, RoomType.WaitingRoom);
+                        break;
+                }
+                
+                return Ok();
+            }
+            catch (KinlyApiException ex)
+            {
+                _logger.LogError(ex, $"Unable to end video hearing {conferenceId}");
```
Comment (levibutler) [inline VideoAPI/Video.API/Controllers/ConferenceManagementController.cs:137]: why is it ending?

File: VideoAPI/Video.API/Controllers/ConferenceManagementController.cs [hunk 3]
```diff
@@ -97,5 +98,45 @@ public async Task<IActionResult> EndVideoHearingAsync(Guid conferenceId)
...
+                        await _videoPlatformService.TransferParticipantAsync(conferenceId,
+                            transferRequest.ParticipantId,
+                            RoomType.HearingRoom, RoomType.WaitingRoom);
+                        break;
+                }
+                
+                return Ok();
+            }
+            catch (KinlyApiException ex)
+            {
```
Comment (levibutler) [inline VideoAPI/Video.API/Controllers/ConferenceManagementController.cs:136]: if the endpoint is called and the user is already in the destination room (already been called) is an exception thrown? what type?

File: VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            var typedResult = (OkResult) result;
+            typedResult.Should().NotBeNull();
+            typedResult.StatusCode.Should().Be((int) HttpStatusCode.OK);
+
+            VideoPlatformServiceMock.Verify(
+                x => x.TransferParticipantAsync(conferenceId, request.ParticipantId, RoomType.HearingRoom,
+                    RoomType.WaitingRoom), Times.Once);
+        }
+        
+        [Test] public async Task should_return_kinly_status_code_on_error()
```
Comment (levibutler) [inline VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs:56]: \n

File: VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            var conferenceId = Guid.NewGuid();
+            var request = new TransferParticipantRequest
+            {
+                ParticipantId = Guid.NewGuid(),
+                TransferType = TransferType.Call
+            };
+            var message = "Auto Test Error";
+            var response = "You're not allowed to start this hearing";
```
Comment (levibutler) [inline VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs:65]: wrong error? doesnt matter i guess but would make it easier to understand the test case if it was relevant maybe?

File: VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            var statusCode = (int) HttpStatusCode.Unauthorized;
+            var exception =
+                new KinlyApiException(message, statusCode, response, null, null);
+            VideoPlatformServiceMock
+                .Setup(x => x.TransferParticipantAsync(It.IsAny<Guid>(), It.IsAny<Guid>(), It.IsAny<RoomType>(),
+                    It.IsAny<RoomType>())).ThrowsAsync(exception);
+            
+            var result = await Controller.TransferParticipantAsync(conferenceId, request);
+            var typedResult = (ObjectResult) result;
```
Comment (levibutler) [inline VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs:74]: we could assert type before casting for a slightly more friendly message
`result.Should().BeOfType<ObjectResult>();`

File: VideoAPI/Video.API/Controllers/ConferenceManagementController.cs [hunk 7]
```diff
@@ -97,5 +98,45 @@ public async Task<IActionResult> EndVideoHearingAsync(Guid conferenceId)
...
+                            RoomType.WaitingRoom, RoomType.HearingRoom);
+                        break;
+                    case TransferType.Dismiss:
+                        _logger.LogDebug(
+                            $"Transferring participant {transferRequest.ParticipantId} out of hearing room");
+                        await _videoPlatformService.TransferParticipantAsync(conferenceId,
+                            transferRequest.ParticipantId,
+                            RoomType.HearingRoom, RoomType.WaitingRoom);
+                        break;
+                }
```
Comment (levibutler) [inline VideoAPI/Video.API/Controllers/ConferenceManagementController.cs:131]: not sure if youll agree but I always like to add a default block and throw an `InvalidOperationException`, I know currently it's not needed but if another type is added sometime
up to you 👍

**Review by levibutler — CHANGES_REQUESTED** [review]
A couple of comments but nice!! need to find you a hobby
🥇

File: VideoAPI/Video.API/Controllers/ConferenceManagementController.cs (hunk 3 above)
Comment (shaed-parkar) [reply VideoAPI/Video.API/Controllers/ConferenceManagementController.cs:136]: No exception thrown. Even invalid ids can be passed. Kinly will return Accepted unless there is an obvious error. Unfortunately even invalid property names are not detected so we need to follow the swagger document closely found [here](https://test.hearings.hmcts.net/swagger-ui/index.html?configUrl=/api/docs/swagger-config#/Hearing/transfer).

NOTE: the API calls actually return 202 despite the document stating it returns OK

File: VideoAPI/Video.API/Controllers/ConferenceManagementController.cs (hunk 2 above)
Comment (shaed-parkar) [reply VideoAPI/Video.API/Controllers/ConferenceManagementController.cs:137]: Typo in the log. Now corrected.

File: VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs (hunk 5 above)
Comment (shaed-parkar) [reply VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs:65]: Updated. Thanks this was a poor copy and paste job

File: VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs (hunk 6 above)
Comment (shaed-parkar) [reply VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs:74]: good call!

File: VideoAPI/Video.API/Controllers/ConferenceManagementController.cs (hunk 7 above)
Comment (shaed-parkar) [reply VideoAPI/Video.API/Controllers/ConferenceManagementController.cs:131]: I've got no objection. Good idea to harden the endpoint and protect ourselves from future unhandled changes

File: VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs (hunk 4 above)
Comment (shaed-parkar) [reply VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs:56]: sorted. thanks

File: VideoAPI/Video.API/Controllers/ConferenceManagementController.cs (hunk 1 above)
Comment (shaed-parkar) [reply VideoAPI/Video.API/Controllers/ConferenceManagementController.cs]: i've re-written the logs to take advantage of custom dimensions now. thank you

File: VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            var request = new TransferParticipantRequest
+            {
+                ParticipantId = Guid.NewGuid(),
+                TransferType = null
+            };
+
+            Assert.ThrowsAsync<InvalidOperationException>(() =>
+                Controller.TransferParticipantAsync(conferenceId, request));
+            
+            // var result = await ;
```
Comment (levibutler) [inline VideoAPI/VideoApi.UnitTests/Controllers/ConferenceManagement/TransferParticipantTests.cs:65]: oop, one final comment :)

**Review by levibutler — APPROVED** [review]

**Review by levibutler — APPROVED** [review]

**Pull request merged**